rarible-subgraph/src/utils.ts

```diff
@@ -151,4 +158,1262 @@ export function getFeeBeneficiaryDetails(
+  proxy: Address;
+  from: Address;
+  constructor(asset: LibAsset, payouts: LibPart[], originFees: LibPart[], proxy: Address, from: Address) {
+    this.asset = asset;
+    this.payouts = payouts;
+    this.originFees = originFees;
+    this.proxy = proxy;
+    this.from = from;
+  }
+}
+
+export class LibAssetType {
+  assetClass: Bytes;
+  data: Bytes;
+  constructor(assetClass: Bytes, data: Bytes) {
+    this.assetClass = assetClass;
+    this.data = data;
+  }
+}
+
+export class LibAsset {
+  assetType: LibAssetType;
+  value: BigInt;
+  constructor(assetType: LibAssetType, value: BigInt) {
+    this.assetType = assetType;
+    this.value = value;
+  }
+}
+
+export function convertAssetToLibAsset(asset: ethereum.Tuple): LibAsset {
+  let tuple = asset;
+  let assetType = tuple[0].toTuple();
+  let value = tuple[1].toBigInt();
+  let assetClass = assetType[0].toBytes();
+  let data = assetType[1].toBytes();
+  return new LibAsset(new LibAssetType(assetClass, data), value);
+}
+
+enum FeeSide {
+  NONE,
+  LEFT,
+  RIGHT,
+}
+
+class LibDealData {
+  maxFeesBasePoint: BigInt;
+  feeSide: FeeSide;
+
+  constructor(maxFeesBasePoint: BigInt, feeSide: FeeSide) {
+    this.maxFeesBasePoint = maxFeesBasePoint;
+    this.feeSide = feeSide;
+  }
+}
+
+export class LibOrder {
+  maker: Address;
+  makeAsset: LibAsset;
+  taker: Address;
+  takeAsset: LibAsset;
+  salt: BigInt;
+  start: BigInt;
+  end: BigInt;
+  dataType: Bytes;
+  data: Bytes;
+
+  constructor(maker: Address, makeAsset: LibAsset, taker: Address, takeAsset: LibAsset, salt: BigInt, start: BigInt, end: BigInt, dataType: Bytes, data: Bytes) {
+    this.maker = maker;
+    this.makeAsset = makeAsset;
+    this.taker = taker;
+    this.takeAsset = takeAsset;
+    this.salt = salt;
+    this.start = start;
+    this.end = end;
+    this.dataType = dataType;
+    this.data = data;
+  }
+}
+
+class LibOrderGenericData {
+  payouts: LibPart[];
+  originFees: LibPart[];
+  isMakeFill: bool;
+  maxFeesBasePoint: BigInt;
+  constructor(payouts: LibPart[], originFees: LibPart[], isMakeFill: bool, maxFeesBasePoint: BigInt) {
+    this.payouts = payouts;
+    this.originFees = originFees;
+    this.isMakeFill = isMakeFill;
+    this.maxFeesBasePoint = maxFeesBasePoint;
+  }
+}
+
+class LibFillResult {
+  leftValue: BigInt;
+  rightValue: BigInt;
+  constructor(leftValue: BigInt, rightValue: BigInt) {
+    this.leftValue = leftValue;
+    this.rightValue = rightValue;
+  }
+}
+
+function calculateTotalAmount(
+  amount: BigInt,
+  orderOriginFees: LibPart[],
+  maxFeesBasePoint: BigInt,
+): BigInt {
+  if (maxFeesBasePoint > BIGINT_ZERO) {
+    return amount;
+  }
+  let total = amount;
+  for (let i = 0; i < orderOriginFees.length; i++) {
+    total = total.plus(bp(amount, orderOriginFees[i].value));
+  }
+  return total;
+}
+
+class doTransfersClass {
+  totalLeftValue: BigInt;
+  totalRightValue: BigInt;
+  royalty: LibPart;
+  originFee: LibPart;
+  payment: BigInt;
+}
+
+function doTransfers(
```

Comment:
It seems this logic is fully duplicate of contract. It better if you can translate the math path to subgraph. Make it more readable & also document it.